File: src/problems/leetcode-cn/712. 两个字符串的最小ASCII删除和/solution/go/solution.go
func minimumDeleteSum(s1 string, s2 string) int {
    rows, cols := len(s1), len(s2)

    dp := make([][]int, rows + 1)
    for i := 0; i <= rows; i++ {
        dp[i] = make([]int, cols + 1)
    }

    for i := 1; i <= rows; i++ {
        dp[i][0] = dp[i-1][0] + int(s1[i-1])
    }
    for i := 1; i <= cols; i++ {
        dp[0][i] = dp[0][i-1] + int(s2[i-1])
    }

    for r := 1; r <= rows; r++ {
        for c := 1; c <= cols; c++ {
            if s1[r-1] != s2[c-1] {
                dp[r][c] = min(dp[r-1][c] + int(s1[r-1]), dp[r][c-1] + int(s2[c-1]))
            } else {
                dp[r][c] = dp[r-1][c-1]
            }
        }
    }
    
    return dp[rows][cols]
}

func min(arr ...int) int {
    length := len(arr)
    if length == 0 {
        return -1
    }
    r := arr[0]
    for i := 1; i < length; i++ {
        if arr[i] < r {
            r = arr[i]
        }
    }
    return r
}

File: src/problems/leetcode-cn/712. 两个字符串的最小ASCII删除和/solution/java/Solution.java
class Solution {
    public int minimumDeleteSum(String s1, String s2) {
        int rows = s1.length(), cols = s2.length();

        int[][] dp = new int[rows+1][];
        for (int i = 0; i <= rows; i++) {
            dp[i] = new int[cols+1];
        }

        for (int i = 1; i <= rows; i++) {
            dp[i][0] = dp[i-1][0] + s1.charAt(i-1);
        }
        for (int i = 1; i <= cols; i++) {
            dp[0][i] = dp[0][i-1] + s2.charAt(i-1);
        }

        for (int r = 1; r <= rows; r++) {
            for (int c = 1; c <= cols; c++) {
                if (s1.charAt(r-1) != s2.charAt(c-1)) {
                    int l = dp[r-1][c] + s1.charAt(r-1), u = dp[r][c-1] + s2.charAt(c-1);
                    dp[r][c] = l < u ? l : u;
                } else {
                    dp[r][c] = dp[r-1][c-1];
                }
            }
        }

        return dp[rows][cols];
    }
}